File: Algorithms/Python3.x/73-Set_Matrix_Zeroes.py
# logistic
# Runtime: 96 ms, faster than 79.75% of Python3 online submissions for Set Matrix Zeroes.
# Memory Usage: 13.4 MB, less than 77.35% of Python3 online submissions for Set Matrix Zeroes.
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        zero_idx = []
        len_row, len_col = len(matrix), len(matrix[0])
        for i, line in enumerate(matrix):
            for j, val in enumerate(line):
                if val == 0: zero_idx.append((i, j))
        for _, (i, j) in enumerate(zero_idx):
            for z in range(len_col):
                matrix[i][z] = 0
            for z in range(len_row):
                matrix[z][j] = 0

File: Algorithms/C++/73-Set_Matrix_Zeroes.cpp
// Runtime: 56 ms, faster than 8.49% of C++ online submissions for Set Matrix Zeroes.
// Memory Usage: 11.4 MB, less than 100.00% of C++ online submissions for Set Matrix Zeroes.
// FIXME:-1000000 as flag is error
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        for (int i = 0; i < matrix.size(); ++i)
            for (int j = 0; j < matrix[0].size(); ++j)
                if (matrix[i][j] == 0) {
                    for (int k = 0; k < matrix[0].size(); ++k) // cur row
                        if (matrix[i][k] != 0)
                            matrix[i][k] = -1000000;
                    for (int k = 0; k < matrix.size(); ++k) // cur col
                        if (matrix[k][j] != 0)
                            matrix[k][j] = -1000000;
                }
        for (int i = 0; i < matrix.size(); ++i)
            for (int j = 0; j < matrix[0].size(); ++j)
                if (matrix[i][j] == -1000000)
                    matrix[i][j] = 0;
    }
};
